_posts/2020-05-05-churn-vs-complexity-vs-coverage.markdown

```diff
@@ -0,0 +1,252 @@
+messages between each other, which will increase coupling and improve cohesion
+per module.
+
+As you go about submitting pull request to ship new features, change existing 
+ones, or patch bugs, the maintainability of your project will vary according
+to the modules you introduce or change.
+
+The bigger the project, the more complexity, the harder it will be to maintain it, 
+because there will be more moving pieces and loosely coupled modules.
+
+If you refactor a huge module, into two smaller modules, you will notice that
+[cognitive complexity](https://docs.codeclimate.com/docs/cognitive-complexity)
+per module will (hopefully) get better. You won't have one huge, complex module 
+anymore, you will have two or more modules that are easier to understand.
+
+Modules that have one and only one responsibility will be easier to maintain 
+and test. Future changes will be easier because you will be able to quickly 
+understand what the module is doing.
+
+In the next section I will explain how you can calculate complexity for all 
+your Ruby modules.
+
+<h3 id="calculating-complexity">Calculating Complexity in Ruby</h3>
+
+There are many tools out there (e.g. [three awesome libraries to assess code quality](https://www.fastruby.io/blog/ruby/quality/code-quality-ruby-gems.html))
+that can help you find the most complex files in your project. One of the most 
+popular tools in the Ruby world is [`flog`](https://github.com/seattlerb/flog).
+
+[Attractor](https://github.com/julianrubisch/attractor) uses `flog` to calculate 
+complexity for each module. I really like the report that it generates. You can 
+see all the modules in one page:
+
+<img src="/blog/assets/images/attractor.png" alt="Attractor Report for MetricFu Ruby gem">
+
+By itself this metric can take you in the wrong direction. Sometimes you will
+find extremely complex files which never change. So they are likely not your
+priority when it's time to refactor code.
+
+<h2 id="churn-vs-complexity">Churn vs. Complexity</h2>
+
+In his article, Michael Feathers, states the following: 
+
+<img src="/blog/assets/images/churn-vs-complexity-graph-by-michael-feathers.jpg" alt="Churn vs Complexity by Michael Feathers">
+
+> These diagrams give us quite a bit of information. The upper right quadrant is 
+particularly important. These files have a high degree of complexity, and they 
+change quite frequently. [...] These classes are particularly ripe for a 
+refactoring investment
+
+I agree that this graph is great for finding _refactoring candidates_. The 
+problem is that sometimes these candidates have no tests to verify their behavior.
+So sometimes we don't have time to:
+
+- Write tests
+- Refactor complex files
+
+Sometimes we have to pick one or the other. So in my opinion it would be great
+to include another metric into our analysis: Code Coverage.
+
+<h2 id="code-coverage">Code Coverage</h2>
+
+I believe the churn vs. complexity graph could be enhanced by adding a new 
```

Comment:
this and the previous paragraph feels repetitive, both say that the previous report can be better if the code coverage metric is added